File: frameworks/oxpecker/Services/Dataset.fs
/// Serves the preloaded JSON dataset for the /json workload. The file is read
/// once, when this module is first touched.
module HttpArena.Services.Dataset

open System
open System.IO
open System.Text.Json

open HttpArena

let private items =
    let path =
        match Environment.GetEnvironmentVariable "DATASET_PATH" with
        | null | "" -> "/data/dataset.json"
        | value -> value

    if File.Exists path then
        JsonSerializer.Deserialize<Item[]>(File.ReadAllText path, Serialization.options)
    else
        null

let getItems (count: int) (multiplier: int) =
    let count = Math.Clamp(count, 0, items.Length)
    let processed = Array.zeroCreate<ProcessedItem> count

    for i in 0 .. count - 1 do
        let item = items[i]
        processed[i] <- {
            Id = item.Id
            Name = item.Name
            Category = item.Category
            Price = item.Price
            Quantity = item.Quantity
            Active = item.Active
            Tags = item.Tags
            Rating = item.Rating
            Total = item.Price * item.Quantity * multiplier
        }

    { JsonResponse.Items = processed; Count = count }

File: frameworks/carter/Dataset.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace HttpArena.Carter;

public sealed class Rating
{
    [JsonPropertyName("score")] public long Score { get; set; }
    [JsonPropertyName("count")] public long Count { get; set; }
}

public sealed class DatasetItem
{
    [JsonPropertyName("id")] public long Id { get; set; }
    [JsonPropertyName("name")] public string Name { get; set; } = "";
    [JsonPropertyName("category")] public string Category { get; set; } = "";
    [JsonPropertyName("price")] public long Price { get; set; }
    [JsonPropertyName("quantity")] public long Quantity { get; set; }
    [JsonPropertyName("active")] public bool Active { get; set; }
    [JsonPropertyName("tags")] public string[] Tags { get; set; } = [];
    [JsonPropertyName("rating")] public Rating Rating { get; set; } = new();
}

// Property order is the wire order: id..rating then the computed total.
public sealed class OutItem
{
    [JsonPropertyName("id")] public long Id { get; set; }
    [JsonPropertyName("name")] public string Name { get; set; } = "";
    [JsonPropertyName("category")] public string Category { get; set; } = "";
    [JsonPropertyName("price")] public long Price { get; set; }
    [JsonPropertyName("quantity")] public long Quantity { get; set; }
    [JsonPropertyName("active")] public bool Active { get; set; }
    [JsonPropertyName("tags")] public string[] Tags { get; set; } = [];
    [JsonPropertyName("rating")] public Rating Rating { get; set; } = new();
    [JsonPropertyName("total")] public long Total { get; set; }
}

public sealed class OutList
{
    [JsonPropertyName("items")] public List<OutItem> Items { get; set; } = [];
    [JsonPropertyName("count")] public int Count { get; set; }
}

/// The dataset, read once at startup. A missing or broken file is not fatal:
/// /json then answers with an empty list.
public sealed class DatasetService
{
    public DatasetItem[] Items { get; }

    public DatasetService()
    {
        var path = Environment.GetEnvironmentVariable("DATASET_PATH") ?? "/data/dataset.json";
        try
        {
            Items = JsonSerializer.Deserialize<DatasetItem[]>(File.ReadAllBytes(path)) ?? [];
        }
        catch
        {
            Items = [];
        }
    }
}
